SearchEngine/lambda_SearchEngine/include/log4pp.h

```diff
@@ -0,0 +1,121 @@
+#pragma once
+#include <iostream>
+#include <string>
+#include <log4cpp/SimpleLayout.hh>
+#include <log4cpp/BasicLayout.hh>
+#include <log4cpp/PatternLayout.hh>
+#include <log4cpp/OstreamAppender.hh>
+#include <log4cpp/FileAppender.hh>
+#include <log4cpp/RollingFileAppender.hh>
+#include <log4cpp/Category.hh>
+#include <log4cpp/Priority.hh>
+
+using namespace std;
+#define suffix(msg) std::string(msg).append(" ##")\
+        .append(__FILE__).append(":").append(__func__)\
+        .append(":").append(int2string(__LINE__))\
+        .append("##").c_str()
+
+#define logError(msg, ...) Mylogger::getInstance()->error(suffix(msg),##__VA_ARGS__)
+#define logWarn(msg, ...) Mylogger::getInstance()->warn(suffix(msg),##__VA_ARGS__)
+#define logInfo(msg, ...) Mylogger::getInstance()->info(suffix(msg),##__VA_ARGS__)
+#define logDebug(msg, ...) Mylogger::getInstance()->debug(suffix(msg),##__VA_ARGS__)
+
+inline std::string int2string(int line) {
+    std::ostringstream oss;
+    oss << line;
+    return oss.str();
+}
+
+using std::cout;
+using std::endl;
+using namespace log4cpp;
+
+
+class Mylogger
+{
+public:
+
+    template<class...Args>
+	void warn(const char * msg,Args...args);
+
+    template<class...Args>
+	void error(const char * msg,Args...args);
+
+    template<class...Args>
+	void debug(const char * msg,Args...args);
+
+    template<class...Args>
+    void info(const char * msg,Args...args);
+    
+
+    static Mylogger * getInstance(){
+        if (_mylogger == nullptr){
+            _mylogger = new Mylogger();
+        }
+        return _mylogger;
+    }
+
+    static void destroy(){
+        if (_mylogger){
+            _mylogger->_Mylog.info("Mylogger destroy");
+            _mylogger->_Mylog.shutdown();
+            delete(_mylogger);
+            _mylogger = nullptr;
+        }
+    }
+private:
+    Mylogger()
+    :_Mylog(Category::getRoot())
+    {  
+        //1、设置日志格式
+        PatternLayout *ptn1 = new PatternLayout;
```

Comment:
we should not handle pointer manually, 
it is a burder for programs to maintain the heap memory.
just use smart pointer
i think we can samply use `std::unique_ptr` here